### PR TITLE
Fix for OUTPUT INSERTED and computed columns

### PR DESCRIPTION
**Pull request by darkdef**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Tests pass?   | ✔️
| Fixed issues  | #18287 
Added key to computed column in table schema and usage for generate query (OUTPUT INSERTED without computed columns)

### CODE REVIEW

File: framework/db/mssql/QueryBuilder.php [hunk 1]
```diff
@@ -475,22 +475,29 @@ public function insert($table, $columns, &$params)
...
-            . ($version2005orLater ? ' OUTPUT INSERTED.* INTO @temporary_inserted' : '')
-            . (!empty($placeholders) ? ' VALUES (' . implode(', ', $placeholders) . ')' : $values);
-
         if ($version2005orLater) {
             $schema = $this->db->getTableSchema($table);
             $cols = [];
+            $columns = [];
             foreach ($schema->columns as $column) {
+                if ($column->isComputed)
+                    continue;
```
Comment (bizley) [inline framework/db/mssql/QueryBuilder.php:484]:
```suggestion
                if ($column->isComputed) {
                    continue;
                }
```

File: framework/db/mssql/QueryBuilder.php [hunk 2]
```diff
@@ -475,22 +475,29 @@ public function insert($table, $columns, &$params)
...
                     . (in_array($column->dbType, ['char', 'varchar', 'nchar', 'nvarchar', 'binary', 'varbinary']) ? "(MAX)" : "")
                     . ' ' . ($column->allowNull ? "NULL" : "");
+                $columns[] = 'INSERTED.' . $column->name;
             }
-            $sql = "SET NOCOUNT ON;DECLARE @temporary_inserted TABLE (" . implode(", ", $cols) . ");" . $sql . ";SELECT * FROM @temporary_inserted";
+        }
+
+        $sql = 'INSERT INTO ' . $this->db->quoteTableName($table)
+            . (!empty($names) ? ' (' . implode(', ', $names) . ')' : '')
+            . (($version2005orLater and count($columns)) ? ' OUTPUT ' . implode(',', $columns) . ' INTO @temporary_inserted' : '')
```
Comment (bizley) [inline framework/db/mssql/QueryBuilder.php:495]:
```suggestion
            . (($version2005orLater && count($columns)) ? ' OUTPUT ' . implode(',', $columns) . ' INTO @temporary_inserted' : '')
```

Comment (bizley) [inline framework/db/mssql/QueryBuilder.php:495]: Also `count($columns)` is used more than once, it's worth to make it a variable.

File: framework/db/mssql/QueryBuilder.php [hunk 3]
```diff
@@ -475,22 +475,29 @@ public function insert($table, $columns, &$params)
...
+            . (!empty($placeholders) ? ' VALUES (' . implode(', ', $placeholders) . ')' : $values);
+
+        if ($version2005orLater and count($columns)) {
```
Comment (bizley) [inline framework/db/mssql/QueryBuilder.php:498]:
```suggestion
        if ($version2005orLater && count($columns)) {
```

File: framework/db/mssql/QueryBuilder.php [hunk 4]
```diff
@@ -475,22 +475,29 @@ public function insert($table, $columns, &$params)
...
+            $sql = "SET NOCOUNT ON;DECLARE @temporary_inserted TABLE (" . implode(", ", $cols) . ");" . $sql .
```
Comment (bizley) [inline framework/db/mssql/QueryBuilder.php:499]:
```suggestion
            $sql = 'SET NOCOUNT ON;DECLARE @temporary_inserted TABLE (' . implode(', ', $cols) . ');' . $sql .
```

File: framework/db/mssql/QueryBuilder.php [hunk 5]
```diff
@@ -475,22 +475,29 @@ public function insert($table, $columns, &$params)
...
+                ";SELECT * FROM @temporary_inserted";
```
Comment (bizley) [inline framework/db/mssql/QueryBuilder.php:500]:
```suggestion
                ';SELECT * FROM @temporary_inserted';
```

File: framework/db/mssql/ColumnSchema.php [hunk 6]
```diff
@@ -14,6 +14,11 @@
  */
 class ColumnSchema extends \yii\db\ColumnSchema
 {
+    /**
+     * @var bool whether this column is a computed column
```
Comment (bizley) [inline framework/db/mssql/ColumnSchema.php:18]: Requires `@since`

**Review by bizley — CHANGES_REQUESTED** [review]

**Review by bizley — APPROVED** [review]

**Review by samdark — APPROVED** [review]
Looks good 👍 Please add a line for CHANGELOG.md. Thanks.

**Comment by samdark** [comment]
👍

**Pull request merged**